azure-communication-ui/chat/src/main/java/com/azure/android/communication/ui/chat/service/sdk/ChatPullEventHandler.kt

```diff
@@ -0,0 +1,151 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.android.communication.ui.chat.service.sdk
+
+import com.azure.android.communication.chat.ChatThreadClient
+import com.azure.android.communication.chat.models.ChatMessageType
+import com.azure.android.communication.chat.models.ChatEventType
+import com.azure.android.communication.chat.models.ListChatMessagesOptions
+import com.azure.android.communication.ui.chat.models.into
+import com.azure.android.communication.ui.chat.models.ChatEventModel
+import com.azure.android.communication.ui.chat.models.ChatThreadInfoModel
+import com.azure.android.communication.ui.chat.models.RemoteParticipantInfoModel
+import com.azure.android.communication.ui.chat.models.RemoteParticipantsInfoModel
+import com.azure.android.communication.ui.chat.service.sdk.wrapper.into
+import com.azure.android.communication.ui.chat.utilities.CoroutineContextProvider
+import com.azure.android.core.util.RequestContext
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+import kotlinx.coroutines.asCoroutineDispatcher
+import org.threeten.bp.OffsetDateTime
+import java.util.concurrent.Executors
+
+internal class ChatPullEventHandler(coroutineContextProvider: CoroutineContextProvider) {
+
+    private val coroutineScope = CoroutineScope((coroutineContextProvider.Default))
+    private val singleThreadedContext = Executors.newSingleThreadExecutor()
+
+    private lateinit var eventSubscriber: (ChatEventModel) -> Unit
+    private lateinit var chatThreadClient: ChatThreadClient
+
+    private val listChatMessagesOptions = ListChatMessagesOptions().apply {
+        maxPageSize = ChatSDKWrapper.PAGE_MESSAGES_SIZE
+    }
+
+    fun fetchMessages(from: OffsetDateTime?) {
+        coroutineScope.launch {
+            withContext(singleThreadedContext.asCoroutineDispatcher()) {
+                fetchMessagesList(from)
+            }
+        }
+    }
+
+    fun start(
+        chatThreadClient: ChatThreadClient,
+        eventSubscriber: (ChatEventModel) -> Unit
+    ) {
+        this.chatThreadClient = chatThreadClient
+        this.eventSubscriber = eventSubscriber
+    }
+
+    fun stop() {
+        singleThreadedContext.shutdown()
+        coroutineScope.cancel()
+    }
+
+    private fun fetchMessagesList(from: OffsetDateTime?) {
```

Comment:
Long method, if possible refactor to functions for readability.

Reply:
resolving as per discussion